src/main/java/de/tum/in/www1/exerciseapp/web/rest/QuizSubmissionResource.java

```diff
@@ -270,24 +272,35 @@ private QuizSubmission submitSubmission(Participation participation, QuizSubmiss
             result = resultRepository.save(result);
             // get previous Result
             Result previousResult = getPreviousResult(result);
-            // add the new Result to the quizPointStatistic and remove the previous one
-            if(previousResult != null) {
-                ((QuizExercise) previousResult.getParticipation().getExercise()).getQuizPointStatistic().removeOldResult(previousResult.getScore(),true);
-            }
-            ((QuizExercise) result.getParticipation().getExercise()).getQuizPointStatistic().addResult(result.getScore(),true);
-            quizPointStatisticRepository.save(((QuizExercise) result.getParticipation().getExercise()).getQuizPointStatistic());
-            // remove the previous Result from the QuestionStatistics
-            if(previousResult != null) {
-                for(SubmittedAnswer submittedAnswer: ((QuizSubmission)previousResult.getSubmission()).getSubmittedAnswers()) {
-                    submittedAnswer.getQuestion().getQuestionStatistic().removeOldResult(submittedAnswer, true);
-                    questionStatisticRepository.save(submittedAnswer.getQuestion().getQuestionStatistic());
+
+            // critical part locked with Semaphore statisticSemaphore
+            try {
+                statisticSemaphore.acquire();
+                QuizExercise quiz = quizExerciseRepository.findOne(participation.getExercise().getId());
+
+                for (Question question: quiz.getQuestions()) {
+                    if(previousResult != null) {
+                        // remove the previous Result from the QuestionStatistics
+                        question.getQuestionStatistic().removeOldResult(((QuizSubmission)previousResult.getSubmission()).getSubmittedAnswerForQuestion(question), true);
+                    }
+                    // add the new Result to QuestionStatistics
+                    question.getQuestionStatistic().addResult(quizSubmission.getSubmittedAnswerForQuestion(question), true);
+                    questionStatisticRepository.save(question.getQuestionStatistic());
                 }
+
+                // add the new Result to the quizPointStatistic and remove the previous one
+                if (previousResult != null) {
+                    quiz.getQuizPointStatistic().removeOldResult(previousResult.getScore(), true);
+                }
+                quiz.getQuizPointStatistic().addResult(result.getScore(), true);
+                quizPointStatisticRepository.save(quiz.getQuizPointStatistic());
+
+            } catch (InterruptedException e) {
+                //TO-DO
```

Comment:
Implement the TO-Do and please provide a meaningful log message in this case. Do you need to report the problem to the client?